addons/stock/models/stock_rule.py

```diff
@@ -101,6 +101,11 @@ def _onchange_route(self):
             self.company_id = self.route_id.company_id
         if self.picking_type_id.warehouse_id.company_id != self.route_id.company_id:
             self.picking_type_id = False
+        if self.company_id:
+            if self.location_id.company_id not in [False, self.company_id]:
+                self.location_id = False
+            if self.picking_type_id.company_id not in [False, self.company_id]:
+                self.picking_type_id = False
```

Comment:
```suggestion
            if self.location_id.company_id.id not in [False, self.company_id.id]:
                self.location_id = False
            if self.picking_type_id.company_id.id not in [False, self.company_id.id]:
                self.picking_type_id = False
```
`record.relational_field` may be falsy, but not equal to `False`, you have to check the id for that.